proposals/0004-FaultTolerance.md

```diff
@@ -0,0 +1,319 @@
+
+    /**
+     * For each retry delay, a randomly portion of the delay multiplied by the jitterFactor will be added or subtracted to the delay.
+     * e.g. a retry delay of 200 milliseconds and a jitter of 0.25 will result in a random retry delay between 150 and 250 milliseconds.
+     * @return the jitter factor.
+     */
+
+    double jitterFactor() default 0.25;
+
+    /**
+     *
+     * @return Specify the failure to retry on
+     */
+    Class<? extends Throwable>[] retryOn() default { Throwable.class };
+
+    /**
+     *
+     * @return Specify the failure to abort on
+     */
+    Class<? extends Throwable>[] aboartOn() default { Throwable.class };
+
+
+}
+
+
+```
+#### CircuitBreaker: a rule to define when to close the circuit
+
+```
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Target({ ElementType.TYPE, ElementType.METHOD })
+public @interface CircuitBreaker {
+
+        /**
+     * Define the failure criteria
+     * @return the failure exception
+     */
+    Class<? extends Throwable>[] failOn() default Throwable.class;
+
+    /**
+     *
+     * @return The delay time after the circuit is open
+     */
+    long delay() default 2;
+
+    /**
+     *
+     * @return The delay unit after the circuit is open
+     */
+
+    TimeUnit delayUnit() default TimeUnit.SECONDS;
+
+    /**
+     *
+     * @return The failure threshold to open the circuit
+     */
+    long failThreshold() default 2;
+
+    /**
+     *
+     * @return The success threshold to fully close the circuit
+     */
+    long successThreshold() default 2;
+
+}
+```
+#### Fallback
+```
+/**
+ * Define the Fallback annotation to specify the fallback callable, BiConsumer or BiFuncation
+ *@author Emily Jiang
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Target({ ElementType.TYPE, ElementType.METHOD })
+public @interface Fallback {
+
+    /**
+     *
+     * @return the fallback class
+     */
+    Class<?> fallback() default Callable.class;
+
+}
+```
+#### Timeout to be used with either Retry or CircuitBreaker
+```
+/**
+ * The Retry annotation to define the number of the retries and the fallback method on reaching the
+ * retry counts.
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ ElementType.METHOD, ElementType.TYPE })
+public @interface TimeOut {
+
+    /**
+     *
+     * @return the timeout
+     */
+    long time() default 2;
+
+    /**
+     *
+     * @return the time unit
+     */
+    TimeUnit timeUnit() default TimeUnit.MILLISECONDS;
+
+}
+```
+#### Bulkhead - threadpool or semaphore style
+```
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+@Target({ ElementType.METHOD, ElementType.TYPE })
+public @interface Bulkhead {
+    public enum Mode {
+        THREADPOOL, SEMAPHORE;
+
+        private Mode() {
+        }
+    }
+
+    Mode value() default Mode.THREADPOOL;
+}
+```
+#### Usage
+An interceptor and fault tolerance policy can be applied to a bean or methods.
+```
+@ApplicationScoped
```

Comment:
This example makes it confusing as to the purpose of this proposal.

Reading this example makes it seem that you're adding retries onto `doWork()` so that anything calling it has that information.

If that's correct, how do we set circuit breakers or other fault tolerant things onto an external call? Does it mean an external call needs to be executed in a method so it can be wrapped with necessary annotations?

I feel some additional examples would help clarify this further